Sofa/Component/Engine/Select/src/sofa/component/engine/select/MeshROI.h

```diff
@@ -32,121 +32,76 @@
 
 protected:
-
     MeshROI();
+    ~MeshROI() override = default;
 
-    ~MeshROI() override {}
 public:
-
-    void init() override;
-    void reinit() override;
-    void doUpdate() override;
-    void draw(const core::visual::VisualParams*) override;
-
-    /// Pre-construction check method called by ObjectFactory.
-    /// Check that DataTypes matches the MechanicalState.
-    template<class T>
-    static bool canCreate(T*& obj, core::objectmodel::BaseContext* context, core::objectmodel::BaseObjectDescription* arg)
-    {
-        if (!arg->getAttribute("template"))
-        {
-            // only check if this template is correct if no template was given
-            if (context->getMechanicalState() && dynamic_cast<sofa::core::behavior::MechanicalState<DataTypes>*>(context->getMechanicalState()) == nullptr)
-            {
-                arg->logError(std::string("No mechanical state with the datatype '") + DataTypes::Name() +
-                              "' found in the context node.");
-                return false; // this template is not the same as the existing MechanicalState
-            }
-        }
-
-        return BaseObject::canCreate(obj, context, arg);
-    }
+    void roiInit() override;
+    bool roiDoUpdate() override;
+    void roiDraw(const core::visual::VisualParams* vparams) override;
 
 protected:
     bool checkSameOrder(const CPos& A, const CPos& B, const CPos& pt, const CPos& norm);
-    bool isPointInMesh(const CPos& p);
+
     bool isPointInIndices(const unsigned int& i);
     bool isPointInBoundingBox(const CPos& p);
-    bool isEdgeInMesh(const Edge& e);
-    bool isTriangleInMesh(const Triangle& t);
-    bool isTetrahedronInMesh(const Tetra& t);
 
+    bool isPointInROI(const CPos& p) override;
+    bool isEdgeInROI(const Edge& e) override;
+    bool isEdgeInStrictROI(const Edge& e) override;
+    bool isTriangleInROI(const Triangle& t) override;
+    bool isTriangleInStrictROI(const Triangle& t) override;
+    bool isQuadInROI(const Quad& q) override;
+    bool isQuadInStrictROI(const Quad& q) override;
+    bool isTetrahedronInROI(const Tetra& t) override;
+    bool isTetrahedronInStrictROI(const Tetra& t) override;
+    bool isHexahedronInROI(const Hexa& t) override;
+    bool isHexahedronInStrictROI(const Hexa& t) override;
 
-    void compute();
+
+protected:
     void checkInputData();
     void computeBoundingBox();
 
 public:
     //Input
-    // Global mesh
-    Data<VecCoord> d_X0; ///< Rest position coordinates of the degrees of freedom
-    Data<type::vector<Edge> > d_edges; ///< Edge Topology
-    Data<type::vector<Triangle> > d_triangles; ///< Triangle Topology
-    Data<type::vector<Tetra> > d_tetrahedra; ///< Tetrahedron Topology
     // ROI mesh
     Data<VecCoord> d_X0_i; ///< ROI position coordinates of the degrees of freedom
     Data<type::vector<Edge> > d_edges_i; ///< ROI Edge Topology
     Data<type::vector<Triangle> > d_triangles_i; ///< ROI Triangle Topology
```

Comment:
What's the difference between those and the one from the base ?

Reply:
One defines the mesh to get indices with, and the other is the shape of the ROI itself (like box or sphere)